Direct shape.byMesh does not support non-planar quads

mjkkirschner opened the pull request:
### Purpose
It was found that the method for converting from Dynamo Meshes to Revit Meshes using the TessellatedShapeBuilder failed when the Dynamo mesh was built with quads and those quads were non planar. The TessellatedShapeBuilder does not support non planar quads, but it does support quads.

~~To fix the issue we look for quad faces in the Dynamo mesh that are non planar by comparing the normals of two triangles defined by quad vertices. If the normals are not parallel than we skip adding the quad and instead add two triangles to the tessellated shape builder that represent this quad.~~

~~We also check for one case where the quad is planar but twisted by checking the quad for self intersections.~~

Instead we just triangulate all quads when converting from a  Proto Mesh to Revit Mesh

http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8322

<img width="986" alt="screen shot 2015-09-10 at 2 28 00 pm" src="https://cloud.githubusercontent.com/assets/508936/9797248/29241f64-57c8-11e5-91f1-106f4448a3a0.png">

![screen shot 2015-09-09 at 4 16 49 pm](https://cloud.githubusercontent.com/assets/508936/9797229/10dea258-57c8-11e5-912e-5f379a7ce5db.png)

<img width="720" alt="screen shot 2015-09-10 at 2 33 11 pm" src="https://cloud.githubusercontent.com/assets/508936/9797400/e91051d0-57c8-11e5-97a2-ddb03b3bdbd4.png">


### Testing
I've added tests for the above pictured cases and a simple non planar quad with one point out of plane.

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate

### Reviewers

@pboyer 

### FYIs
@jnealb 
@Randy-Ma

ikeough commented on src/Libraries/RevitNodes/GeometryConversion/ProtoToRevitMesh.cs:
```diff
@@ -121,7 +140,27 @@ public static class ProtoToRevitMesh
+                    //if the two triangle normals are not parallel then we have a non planar quad
+                    //and we'll skip adding this face and instead add two new faces to list to process
+                    //these new faces are two triangles representing the quad
+                    if (Math.Abs(CABnormal.Dot(DACNormal) - 1) > 0.000001 ||
```
Make this "magic number" `private const readonly precision`.

ikeough commented on src/Libraries/RevitNodes/GeometryConversion/ProtoToRevitMesh.cs:
```diff
@@ -121,7 +140,27 @@ public static class ProtoToRevitMesh
+                        QuadSelfIntersects(currentVerts)
+                        )
+                    {
+                        indicies.Add(IndexGroup.ByIndices(f.B, f.C, f.A));
```
Spelling police: "Indices"

ikeough commented on test/Libraries/RevitNodesTests/Elements/DirectShapeTests.cs:
```diff
@@ -148,6 +156,114 @@ public void ByName_NonexistentName()
+            var p3 = Point.ByCoordinates(5.0, 0.0, 0.0);
+            var p4 = Point.ByCoordinates(0.0, 5.0, 0.0);
+
+            var index1 = IndexGroup.ByIndices(0, 1, 2, 3);
```
Starting here, I see a lot of copy paste code. Refactor the code that is common amongst these test methods. Something like `CreateDirectShapeFromQuadPoints`. It will make these tests easier to read.

ikeough commented:
@mjkkirschner Nice fix. A couple of small comments.

mjkkirschner commented:
@pboyer @ikeough after speaking with pboyer about this fix we decided to instead just triangulate all incoming quads as this is less likely to break in the future as tolerances change in TessellatedShapeBuilder. I've refactored the tests and fixed the typos as well as modified the ProtoToRevitMesh to no longer look for planar quads, instead any quad is triangulated.

jnealb commented:
@mjkkirschner Perhaps put your work in a package?

ikeough commented:
@mjkkirschner Sounds good to me. Thanks! LGTM.

mjkkirschner commented:
@pboyer I'll wait to merge and cherry pick this until you get a chance to take a look.

pboyer commented on src/Libraries/RevitNodes/GeometryConversion/ProtoToRevitMesh.cs:
```diff
@@ -102,35 +101,42 @@ public static class ProtoToRevitMesh
-                    var currentindex = Convert.ToInt32(currentIndicies[i]);
+                    indices.Add(IndexGroup.ByIndices(f.B, f.C, f.A));
+                    indices.Add(IndexGroup.ByIndices(f.A, f.C, f.D));
+                    continue;
```
Hate to be the performance guy, but is there a strong reason to add to the original list while iterating over it? Your iteration will terminate, but the code is a bit strange to read and memory intensive. Consider the case where you're dealing with a 1 million face quad mesh - now you have 3 millions index groups in this list.

pboyer replied:
You should also keep in mind that the `IndexGroup.ByIndices` call actually has a lot of function call overhead.

mjkkirschner commented:
@pboyer @ikeough - I profiled this code on Pboyer's suggestion - and made some changes based on that. I removed a linq call, removed a toList call on the Mesh.indicies, and avoid creating a new list of indicies at any point, instead just adding the faces to the tessellatedShape builder immediately.

@pboyer please take another look.

pboyer commented on src/Libraries/RevitNodes/GeometryConversion/ProtoToRevitMesh.cs:
```diff
@@ -102,35 +101,34 @@ public static class ProtoToRevitMesh
             tsb.OpenConnectedFaceSet(false);
 
-            foreach (var f in indicies)
+            for (int faceindex = 0; faceindex < indices.Count(); faceindex++)
```
If the length of the array is not changing, you can do this to remove the cost of `indices.Count()`:

```csharp
for (var i = 0, l = indices.Count(); i < l; i++){
...
}
```

pboyer commented on src/Libraries/RevitNodes/GeometryConversion/ProtoToRevitMesh.cs:
```diff
@@ -102,35 +101,34 @@ public static class ProtoToRevitMesh
+                else
+                {
+                    AddProtoFaceToTSB(tsb, f, verts, performHostUnitConversion, MaterialId);
                 }
```
Nice, this is what I would've expected.

pboyer commented on src/Libraries/RevitNodes/GeometryConversion/ProtoToRevitMesh.cs:
```diff
@@ -140,5 +138,28 @@ public static class ProtoToRevitMesh
+        //this method converts a proto indexGroup and verts to a tessellated face, and adds it
+        //to the tessellated shape builder that is passed in.
+        //the verts array should be the entire vert array extracted from a proto mesh.
+        private static void AddProtoFaceToTSB(TessellatedShapeBuilder tsb,
```
Use XML documentation with three forward slashes.

pboyer commented:
@mjkkirschner This definitely addresses my concerns. Please take a look at the above comments (both minor) and feel free to merge.
